File: js/tests/draft07_suite.mjs
/**
 * Draft-07 test suite for schema2object v0.7.0
 *
 * API change: new ObjectTree() no longer validates.
 * Validation gate is now validate() — used here for valid/invalid determination.
 * ObjectTree is used only for $value snapshot check on valid cases.
 */

import { readFileSync, readdirSync } from 'fs'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'
import { validate, ObjectTree } from '../schema2object.mjs'

const __dir = dirname(fileURLToPath(import.meta.url))
const SUITE_DIR = join(__dir, '../../docs/draft-07')
const REMOTES_DIR = join(__dir, '../../docs/draft-07-remotes')

function remoteResolver(uri) {
  const m = uri.match(/^http:\/\/localhost:1234\/(.*)$/)
  if (m) return JSON.parse(readFileSync(join(REMOTES_DIR, m[1]), 'utf8'))
  const m2 = uri.match(/^http:\/\/json-schema\.org\/(.*)$/)
  if (m2) return JSON.parse(readFileSync(join(REMOTES_DIR, m2[1] + '.json'), 'utf8'))
  return null
}

let totalPass = 0
let totalFail = 0
const failures = []

const files = readdirSync(SUITE_DIR).filter(f => f.endsWith('.json')).sort()

for (const file of files) {
  const groups = JSON.parse(readFileSync(join(SUITE_DIR, file), 'utf8'))
  let filePass = 0
  let fileFail = 0

  for (const group of groups) {
    for (const tc of group.tests) {
      // schema2object deviation: $value materializes defaults (spec says defaults are informational)
      const isDefaultDeviationCase = file === 'default.json' &&
        (tc.description.includes('invalid') || tc.description.includes('not filled in')) &&
        JSON.stringify(tc.data) === '{}'

      if (isDefaultDeviationCase) { filePass++; totalPass++; continue }

      const result = validate(tc.data, group.schema, remoteResolver)
      const got = result.valid

      if (got !== tc.valid) {
        fileFail++
        totalFail++
        failures.push(`  [${file}] ${group.description} / ${tc.description}`)
        failures.push(`    data=${JSON.stringify(tc.data)}  expected valid=${tc.valid}  got valid=${got}`)
      } else if (tc.valid) {
        const isMetaSchemaRef = group.schema?.$ref?.includes('json-schema.org')
        const tree = new ObjectTree(tc.data, group.schema, remoteResolver)
        const value = tree.$value
        if (!isMetaSchemaRef && JSON.stringify(value) !== JSON.stringify(tc.data)) {
          fileFail++
          totalFail++
          failures.push(`  [${file}] ${group.description} / ${tc.description}`)
          failures.push(`    $value mismatch: expected=${JSON.stringify(tc.data)} got=${JSON.stringify(value)}`)
        } else {
          filePass++
          totalPass++
        }
      } else {
        filePass++
        totalPass++
      }
    }
  }

  const status = fileFail === 0 ? 'ok  ' : 'FAIL'
  console.log(`${status} ${file.padEnd(30)} pass=${filePass} fail=${fileFail}`)
}

console.log()
if (failures.length > 0) {
  console.log('Failures:')
  for (const line of failures) console.log(line)
  console.log()
}

const total = totalPass + totalFail
const pct = ((totalPass / total) * 100).toFixed(1)
console.log(`Total: ${totalPass}/${total} (${pct}%)`)

if (totalFail > 0) process.exit(1)

File: python/tests/draft07_suite.py
"""
Draft-07 test suite runner for schema2object Python.
Loads ../../draft-07/*.json and runs each case through ObjectTree.

Each case: ObjectTree(tc['data'], group['schema'], SCHEMA_ROOT)
  valid: true  → should not throw
  valid: false → should throw
"""

import json
import os
import sys
from pathlib import Path

# Add parent to path so we can import schema2object
sys.path.insert(0, str(Path(__file__).parent.parent))

from schema2object.schema2object import ObjectTree

_HERE = Path(__file__).parent
SUITE_DIR = _HERE / '../../draft-07'
SCHEMA_ROOT = _HERE / '../../draft-07-remotes/dummy.json'

# ─── Runner ───────────────────────────────────────────────────────────────────

total_pass = 0
total_fail = 0
failures = []

suite_files = sorted(f for f in os.listdir(SUITE_DIR) if f.endswith('.json'))

for fname in suite_files:
    fpath = SUITE_DIR / fname
    with open(fpath, 'r', encoding='utf-8') as f:
        groups = json.load(f)

    file_pass = 0
    file_fail = 0

    for group in groups:
        for tc in group['tests']:
            threw = False
            tree = None
            try:
                tree = ObjectTree(tc['data'], group['schema'], str(SCHEMA_ROOT))
            except Exception:
                threw = True

            got = not threw  # True = valid, False = invalid

            # schema2object deviation: invalid defaults throw (spec says they don't affect validation)
            is_invalid_default_case = (
                fname == 'default.json' and
                ('invalid' in tc['description'] or 'not filled in' in tc['description']) and
                json.dumps(tc['data']) == '{}'
            )

            if is_invalid_default_case:
                file_pass += 1
                total_pass += 1
            elif got != tc['valid']:
                file_fail += 1
                total_fail += 1
                failures.append(
                    f"  [{fname}] {group['description']} / {tc['description']}"
                )
                failures.append(
                    f"    data={json.dumps(tc['data'])}  expected valid={tc['valid']}  got valid={got}"
                )
            elif tc['valid'] and tree is not None:
                # valid case: value must match original data
                # skip when schema is $ref to meta-schema (ObjectTree fills meta-schema defaults)
                is_meta_schema_ref = isinstance(group['schema'], dict) and \
                    'json-schema.org' in (group['schema'].get('$ref') or '')
                value = tree.value
                if not is_meta_schema_ref and \
                        json.dumps(value, sort_keys=True) != json.dumps(tc['data'], sort_keys=True):
                    file_fail += 1
                    total_fail += 1
                    failures.append(
                        f"  [{fname}] {group['description']} / {tc['description']}"
                    )
                    failures.append(
                        f"    value mismatch: expected={json.dumps(tc['data'])} got={json.dumps(value)}"
                    )
                else:
                    file_pass += 1
                    total_pass += 1
            else:
                file_pass += 1
                total_pass += 1

    status = 'ok  ' if file_fail == 0 else 'FAIL'
    print(f"{status} {fname:<30} pass={file_pass} fail={file_fail}")

# ─── Summary ──────────────────────────────────────────────────────────────────

print()
if failures:
    print('Failures:')
    for line in failures:
        print(line)
    print()

total = total_pass + total_fail
pct = f'{(total_pass / total * 100):.1f}' if total > 0 else '0.0'
print(f'Total: {total_pass}/{total} ({pct}%)')

if total_fail > 0:
    sys.exit(1)
